### DISCUSSION TITLE
How to send message back to ros 2?

### DISCUSSION THREAD

**Discussion by morkovka1337** [foxglove/community#769] [Q&A]
Hello.
I'm trying to send message back to ros using cyclonedds. My setup is foxbridge & foxglove studio 1.50 (docker image), I'm also using cyclonedds, ros2 galactic and `RMW_IMPLEMENTATION=rmw_fastrtps_cpp`. I've created a Publish panel and set datatype to Float32. 
![image](https://github.com/foxglove/community/assets/32310433/94554118-842d-4fb2-8840-4b896fcdc3fc)
My receiver code is below:
```
import rclpy
from rclpy.node import Node

from std_msgs.msg import Float32


class MinimalSubscriber(Node):

    def __init__(self):
        super().__init__('minimal_subscriber')
        self.subscription = self.create_subscription(
            Float32,
            'upd_speed',
            self.listener_callback,
            10)
        self.subscription  # prevent unused variable warning

    def listener_callback(self, msg):
        print(f'I heard {msg.data}')


def main(args=None):
    rclpy.init(args=args)

    minimal_subscriber = MinimalSubscriber()

    rclpy.spin(minimal_subscriber)

    # Destroy the node explicitly
    # (optional - otherwise it will be done automatically
    # when the garbage collector destroys the node object)
    minimal_subscriber.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()
```

Now when I press the button publish on publish panel, nothing is printed in the receiver. What am I doing wrong? Appreciate any help and comments, thanks!

**Comment by achim-k** [comment]
Does `ros2 topic echo upd_speed` also not show anything when you publish from Studio?

**Reply by morkovka1337** [reply]
Thanks, I will try it

**Comment by defunctzombie** [comment]
Foxglove Studio 1.50 is quite old now - please update to the latest version and try again.

Which version of foxglove-ros-bridge are you using? Make sure you are on the latest version of that as well.

**Reply by morkovka1337** [reply]
I'm using studio 1.50 because when I used the latest version it was not working correctly on iPad apparently

**Reply by defunctzombie** [reply]
Please try the latest version. Also note that we only support Chrome. We do not test or support other browsers so results may vary.

**Reply by morkovka1337** [reply]
As for the version of foxglove-ros-bridge, it is installed from the apt in the docker, I will check the exact version, but I assume it is the latest.

**Comment by ChernyakKonstantin** [comment]
The data can be read with `ros2 topic echo --raw /upd_speed` or with 

```
self.subscription = self.create_subscription(
    Float32,  # Whatever here, due to raw=True below it has no impact.
    '/upd_speed',
    self.listener_callback,
    10,
     raw=True 
)
```

and 

```
def listener_callback(self, msg):
    decoded_msg = msg.decode('utf-8').rstrip('\x00')
    data = json.loads(decoded_msg)
    <YOUR_CODE_HERE_TO_DEAL_WITH_THE_PARSED_DICTIONARY>
```


The problem, as far as I understood, is that the Foxglove Studio sends serialized JSONs those, for unknown reasons, are not proceed correctly.
